File: perl/recur.pl
use strict;

print "Enter the upper bound: ";
my $n = <STDIN>;
my @largest = ([1, 0]);

sub collatz
{
    my $n = $_[0];
    if($n == 1)
    {
        return 0;
    }
    else
    {
        if($n % 2 == 0)
        {
            return collatz($n/2) + 1;
        }
        else
        {
            return collatz(3*$n + 1) + 1;
        }
    }
}

sub c1
{
    return (($a->[1] <=> $b->[1]) or ($b->[0] <=> $a->[0]));
}

sub c2
{
    return (($b->[1] <=> $a->[1]) or ($a->[0] <=> $b->[0]));
}

for(my $i=1; $i <= $n; $i++)
{
    my $l = collatz($i);

    if($l > $largest[0][1])
    {
        if(scalar @largest >= 10)
        {
            shift(@largest);
        }
        push(@largest, [$i, $l]);
        @largest = sort c1 @largest;
    }
}

@largest = sort c2 @largest;
for(my $i=0; $i < scalar @largest; $i++)
{
    my $a = $largest[$i][0];
    my $b = $largest[$i][1];
    print "$a $b\n";
}

File: ruby/recur.rb
print "Enter the upper bound: "
n = STDIN.gets.chomp.to_i
largest = [[1, 0]]

def collatz(n)
    if(n == 1)
        return 0
    else
        if(n % 2 == 0)
            return collatz(n/2) + 1
        else
            return collatz(3*n + 1) + 1
        end
    end
end

for i in 1..n do
    l = collatz(i)
    temp = largest[0][1]
    if l > temp
        if largest.length >= 10
            largest.shift
        end
        e = [i, l]
        largest.push(e)
        largest = largest.sort { |a, b| [a[1], a[0]] <=> [b[1], b[0]] }
    end
end

largest = largest.sort { |a, b| [b[1], a[0]] <=> [a[1], b[0]] }
for e in largest do
    puts "#{e[0]} #{e[1]}"
end
